### PR TITLE
Added backwardsCompatibility flag to AppBarTheme

### LINKED ISSUE #72206: Opt in to use new AppBarTheme features in AppBarTheme, not only in an AppBar
## Use case

The new `AppBarTheme` features implemented via https://github.com/flutter/flutter/pull/71184 currently available on (Channel **master, 1.25.0-9.0.pre.24**) that: 

Fixes #https://github.com/flutter/flutter/issues/70645
Fixes #https://github.com/flutter/flutter/issues/67921
Fixes #https://github.com/flutter/flutter/issues/67497
Fixes #https://github.com/flutter/flutter/issues/61618
Fixes #https://github.com/flutter/flutter/issues/51820
Fixes #https://github.com/flutter/flutter/issues/50606

Allows us to, for example, easily create a white app bar theme for a light themed application. Like shown below, when applied to the Flutter default counter app:

```dart
void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'AppBar Theme Test',
      theme: ThemeData(
        brightness: Brightness.light,
        primarySwatch: Colors.blue,
        appBarTheme: AppBarTheme(
          backgroundColor: Colors.white,
          foregroundColor: Colors.black87,
          elevation: 0,
        ),
      ),
      home: MyHomePage(title: 'AppBar Theme Test'),
      debugShowCheckedModeBanner: false,
    );
  }
}

```

This now gives us a themed application with a white app bar based on an app bar theme, that has dark text, even though the primary color is of brightness dark and our overall theme is of brightness light. Older versions would have forced the text theme to white for the app bar, making it invisible. This is a welcome improvement. It also works if you switch the brightness to Brightness.dark, the app bar keeps it colors as specified via the app bar theme.

image

## Practical adoption issue of the new app bar theme features during migration period

In order to use the new app bar theming features, developers have to opt in to use the new features on an `AppBar()`instance/descendant level, by setting the property `backwardsCompatibility` to `false` (it defaults to `true`), like shown below for the completion of the shown counter app example.

```dart
class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        // Must set to to FALSE to opt in on the new app bar theme features
        backwardsCompatibility: false,
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      drawer: Drawer(),
      endDrawer: Drawer(),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}
```

Currently you have to add the property `backwardsCompatibility` and set it to `false` for every `AppBar` in your app where you want to use the new `AppBar` theming features. This makes opting in to use it cumbersome and it may also slow down migration to the new feature, since you have to introduce and set `backwardsCompatibility` to `false` for every `AppBar()` that you create/have in your app, and there might be quite a large number of them. Later, when the flag is deprecated and totally removed, you have to remove the property again from **all** instances.

To make adoption and migration easier, and to make it simpler to opt in for all app bars in an app, it would be useful to have a way to toggle the desired default behavior via a flag in the `AppBarTheme`. This would be similar to how you can do that with `useTextSelectionTheme` in the overall `ThemeData`.

Maybe I missed something, but I looked in `AppBarTheme` and even `ThemeData`, but I did not see any opt in option to the new app bar theming features before the actual `AppBar()` creation.

To also have a temporary setting, e.g. `useBackwardsCompatibility` property in the `AppBarTheme` would be very useful for adoption and migration. This setting, when defined, would then be used as the default for the app bar's own value, that as now would still have its own property, it is still useful if there is a need to override the theme based setting on an `AppBar` level, perhaps for some migration use cases.

If this option was available, it would be possible to set the property already in the `AppBarTheme()` in the example above. It would then not be necessary to opt in again when you actually create `AppBar()`'s in your application, like in the above example. You could omit it there, since you already opted in via the `AppBarTheme` it inherits its theme from. The theme you probably just define in one place in your app, or in much fewer places than you create app bars at least. This feature would thus make both migration and adoption smoother.

## Proposal

Introduce a new `useBackwardsCompatibility` boolean property in `AppBarTheme`, that when defined controls the default for app bars' `backwardsCompatibility` property in an `AppBar()` that inherits its theme from the app bar theme in question. An `AppBar()` instance can still use its `backwardsCompatibility` property to override the theme inherited value, as is common for many/most theme properties.

It is possible to introduce this property and enhancement without any breaking changes.

Currently the usefulness of theming properties that you can define via a theme, but not actually apply to widgets using the theme, without also setting a property value on widget instances using the theme, is a bit broken imo and partially defeats the purpose of using a theme. 

### Flutter doctor

<details>
  <summary>Flutter doctor output</summary>

```
>>flutter doctor -v
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, 1.25.0-9.0.pre.24, on Microsoft Windows [Version 10.0.18363.1198], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.0)
[√] Android Studio (version 4.1.0)
[√] IntelliJ IDEA Community Edition (version 2020.2)
[√] VS Code (version 1.51.1)
[√] Connected device (4 available)

• No issues found!
PS C:\Users\mryds\OneDrive\code\flutter\_issues\appbar_theme> fvm flutter doctor -v
[√] Flutter (Channel master, 1.25.0-9.0.pre.24, on Microsoft Windows [Version 10.0.18363.1198], locale en-US)
    • Flutter version 1.25.0-9.0.pre.24 at C:\Users\mryds\fvm\versions\master
    • Framework revision a1070e4f4c (35 hours ago), 2020-12-10 23:24:21 -0800
    • Engine revision 827aa5d073
    • Dart version 2.12.0 (build 2.12.0-135.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\mryds\AppData\Local\Android\sdk
    • Platform android-30, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.0)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.8.30709.132
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] IntelliJ IDEA Community Edition (version 2020.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.3.1
    • Flutter plugin version 52.0.4
    • Dart plugin version 202.8070

[√] VS Code (version 1.51.1)
    • VS Code at C:\Users\mryds\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.16.0

[√] Connected device (4 available)
    • SM T510 (mobile)  • R52N906MLAV • android-arm    • Android 10 (API 29)
    • Windows (desktop) • windows     • windows-x64    • Microsoft Windows [Version 10.0.18363.1198]
    • Chrome (web)      • chrome      • web-javascript • Google Chrome 87.0.4280.88
    • Edge (web)        • edge        • web-javascript • Microsoft Edge 87.0.664.57

• No issues found!

```
</details>

---
cc: @HansMuller

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes https://github.com/flutter/flutter/issues/72206

### CODE REVIEW

**Review by darrenaustin — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -609,7 +608,9 @@ class AppBar extends StatefulWidget implements PreferredSizeWidget {
   /// [iconTheme], [actionsIconTheme] properties, and the original use of
   /// the [textTheme] and [brightness] properties.
   ///
-  /// This property is true by default.
+  /// If this property is null, then [AppBarTheme.backwardsCompatibility] of
+  /// [ThemeData.appBarTheme] is used. If that is also null, the default
+  /// value is true.
   ///
   /// This is a temporary property. When setting it to false is no
   /// longer considereed a breaking change, it will be depreacted and
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/app_bar.dart:616]:
```suggestion
  /// longer considered a breaking change, it will be deprecated and
```

Comment (HansMuller) [reply packages/flutter/lib/src/material/app_bar.dart]: Good catch! Thanks.

**Comment by rydmike** [comment]
Nice, thanks @HansMuller ! 💙

**Pull request merged**